ssl/ssl_key_share.cc

```diff
@@ -176,20 +178,550 @@ class X25519KeyShare : public SSLKeyShare {
+                Span<const uint8_t> peer_key) override {
+      if (!out_alert) {
+        OPENSSL_PUT_ERROR(SSL, ERR_R_PASSED_NULL_PARAMETER);
+        return false;
+      }
+
+      // Set alert to internal error by default
+      *out_alert = SSL_AD_INTERNAL_ERROR;
+
+      if (!out_secret || !peer_key.data()) {
+        OPENSSL_PUT_ERROR(SSL, ERR_R_PASSED_NULL_PARAMETER);
+        return false;
+      }
+
+      // hybrid_group_ should have been initialized
+      // in a previous call to Offer()
+      if (!hybrid_group_) {
+        OPENSSL_PUT_ERROR(SSL, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
+        return false;
+      }
+
+      // A hybrid shared secret with two components should be 64 bytes.
+      // If it happens to be larger, the CBB will grow accordingly.
+      CBB hybrid_shared_secret;
+      if (!CBB_init(&hybrid_shared_secret, 64)) {
+        OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+        return false;
+      }
+
+      // Finish() each component key share. Each component's Finish() function
+      // will generate a shared secret. The hybrid shared secret is the
+      // concatenation of all component shared secrets.
+      size_t peer_key_index = 0;
+      for (size_t i = 0; i < NUM_HYBRID_COMPONENTS; i++) {
+        size_t component_key_size = hybrid_group_->accept_share_sizes[i];
+
+        // Verify that |peer_key| contains enough data
+        if (peer_key_index + component_key_size > peer_key.size()) {
+          CBB_cleanup(&hybrid_shared_secret);
+          *out_alert = SSL_AD_DECODE_ERROR;
+          OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+          return false;
+        }
+
+        Span<const uint8_t> component_key =
+          peer_key.subspan(peer_key_index, component_key_size);
+
+        Array<uint8_t> component_secret;
+        if (!key_shares_[i] ||
+            !key_shares_[i]->Finish(&component_secret, out_alert, component_key) ||
+            !CBB_add_bytes(&hybrid_shared_secret, component_secret.data(), component_secret.size())) {
+          CBB_cleanup(&hybrid_shared_secret);
+          OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+          return false;
+        }
+
+        peer_key_index += component_key_size;
+      }
+
+      // Final validation that |peer_key| was the correct size
+      if (peer_key_index != peer_key.size()) {
+        CBB_cleanup(&hybrid_shared_secret);
+        *out_alert = SSL_AD_DECODE_ERROR;
+        OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+        return false;
+      }
+
+      // Retain the hybrid shared secret for our use.
+      if (!CBBFinishArray(&hybrid_shared_secret, out_secret)) {
+        CBB_cleanup(&hybrid_shared_secret);
+        OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+        return false;
+      }
+
+      // Success; clear the alert
+      *out_alert = 0;
+      return true;
+    }
+
+  private:
+   bool init_component_key_shares() {
+      // The component key shares should only be initialized once
```

Comment:
Should this be done in the constructor?

Reply:
+1, moved this function into the constructor